### DISCUSSION TITLE
Multi Pass Shaders

### DISCUSSION THREAD

**Discussion by torinmb** [shader-park/shader-park-core#66] [Ideas]
Add ability to write to multiple buffers. This would allow post-processing + feedback loops.
We'd need to implement this feature in 3 different targets and also ensure it's turned off by default to support older versions.

**Comment by torinmb** [comment]
Potential helper library https://github.com/martinlaxenaire/three-multipass-post-processing. It's minimal, uses MIT license and is inspired by a blog post from a senior engineer from active theory that highlights optimizations for multipass

**Comment by torinmb** [comment]
from pete:
```

renderBufferA( () => {

  let c = enable2D() // really we want fragcoord?
  // let c = vec2("gl_FragCoord.xy", null, true);
  let val = readBufA(texCoord)
  color(val.x, val.y, val.z)
})

let s = getSpace()
let texCoord = vec2(s.x, s.y)
let val = readBufA(texCoord)
box(0.5, 0.5, 0.2+0.1*val)

```

**Reply by torinmb** [reply]
ToDo:

- [x] Expose Sampler2D / Texture in JS
- [ ] implement separatePasses to extract passes from tree and return as separate dictionaries
- [ ] Figure out how to handle buffer initialization (should it be a setup function call where you can write your own SP shader that runs for 1 frame, or is it more helpful to be able to handle a single frame in the body of your buffer code? Eg. there's a boolean exposed  )
- [ ] start to implement multiple buffers in targets 

```
renderBufferA( () => {
  setupBuffer(() => {

  });
  color(val.x, val.y, val.z)
})
```

**Reply by PWhiddy** [reply]
Updated TODO 

- [x] Expose sampler2D/texture in js
- [x] implement a multipass to GLSL func (prepends common tab to other passes and then runs them through sculptToGLSL
- [ ] Resolve the buffer initialization debacle
- [ ] implement targets

**Comment by torinmb** [comment]
Need an option to:
1) set RT wrap mode: between wrap, clear, hold, tile etc.
2) set min + mag filter using LinearFilter, or NearestFilter
